Localize images from Markdown metadata

stasinos opened the pull request:
Read Lat, Lon fields in the Markdown files to localize images. These override the EXIF GPS coordinates, if both are present, allowing manually localizing the images on the map. This is useful to, for instance, localize the image based on what is depicted instead of where the camera was standing.

stasinos committed:
Use markdown metadata to provide GPS

GPS coordinates can be given in Lat,Lon fields
in the Markdown files.
These override the EXIF GPS coordinates, if
both are present.

stasinos committed:
Avoid exception handling

saimn commented on docs/image_information.rst:
```diff
@@ -16,6 +16,10 @@ keys are used by Sigal to get the useful informations on the gallery:
   e.g. for videos and some image formats.
+- *Lat* and *Lon*: geo-location for positioning the image on map galleries.
+  Note that this overrides EXIF coordinates, so except for localizing when
+  EXIF cooredinates are missing it can also be used to localize the image
```
```suggestion
  EXIF coordinates are missing it can also be used to localize the image
```

saimn commented on src/sigal/gallery.py:
```diff
@@ -286,6 +287,15 @@ def _get_markdown_metadata(self):
 
+    @cached_property
+    def lat(self):
+        return self.markdown_metadata.get("lat", {})
```
```suggestion
        return self.markdown_metadata.get("lat", "")
```
should return a string by default I think ?

stasinos replied:
Or None which simplifies the code even further, allowing get() its default behaviour.
The related if statement in Jinja works the same.

saimn commented on src/sigal/gallery.py:
```diff
@@ -286,6 +287,15 @@ def _get_markdown_metadata(self):
+    @cached_property
+    def lon(self):
+        """Other metadata extracted from the Markdown index.md file."""
+        return self.markdown_metadata.get("lon", {})
```
```suggestion
        return self.markdown_metadata.get("lon", "")
```

stasinos replied:
Or None which simplifies the code even further, allowing get() its default behaviour.
The related if statement in Jinja works the same.
Please see new commit.

saimn commented on src/sigal/gallery.py:
```diff
@@ -286,6 +287,15 @@ def _get_markdown_metadata(self):
 
+    @cached_property
+    def lat(self):
+        return self.markdown_metadata.get("lat", {})
```
Can you change the properties to return `media.exif.gps.lat/lon` if not available in meta ? this would avoid the if/else in the template.

stasinos replied:
But that would make it impossible to write a template that prefers the EXIF coordinates.
You might not want this, IMHO.

saimn replied:
people can still use directly `media.exif.gps`. But for `media.lat` I think it makes sense to avoid the manual coordinates or the exif ones.

stasinos replied:
I hate to insist, but I feel I have to. I perceive the Python code as a considerably "deeper" layer than the templates. I expect that many users will be happy to work at the template layer but will want to avoid the Python layer. Thus, I expect that a simple, easily predictable behaviour for media.lat/lon and media.exif.gps makes it easy to write templates.

Now, if you really want to have a template without ifs, I suggest keeping meta.lat/lon as they are and adding a new set of variables that implement the fallback and a configuration option that defines preference to GPS or .md if both are present.

stasinos replied:
And, BTW, come to think of it, one of the ifs will still be needed as it might be the case that there are no coordinates at either EXIF or .md. So we will still need an if, except without the else branch.

stasinos replied:
I just had another idea: How about a config option that sets the order of preference between md, EXIF?
Then both ways are accessible at the template. If the option also includes a "default location" (say, 0,0 or someplace in the middle of the Pacific) then all photos are guaranteed to have a location and the template can be completely without checks. This has the advantage that photos without a location do not silently disappear but are placed at a location that the users know to mean that they have forgotten to add a location.

saimn reviewed (commented):
Seems useful, thanks. A few comments below.

stasinos committed:
Fix typo in docs/image_information.rst

Co-authored-by: Simon Conseil <contact@saimon.org>

stasinos committed:
Change lat, lon properties to return None

Slight code simplification, allowing get() to
return None when the field is missing. The related
if statement in Jinja2 accepts None as False.

Co-authored-by: Simon Conseil <contact@saimon.org>
